File: src/main/java/me/partlysanestudios/partlysaneskies/system/commands/CommandManager.java
//
// Written by Su386.
// See LICENSE for copyright and license notices.
//

package me.partlysanestudios.partlysaneskies.system.commands;

import net.minecraft.command.CommandException;
import net.minecraft.command.ICommand;
import net.minecraft.command.ICommandSender;
import net.minecraft.util.BlockPos;
import net.minecraftforge.client.ClientCommandHandler;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class CommandManager {
    public static HashMap<String, PSSCommand> commandList = new HashMap<>();

    public static ICommand registerCommand(PSSCommand pssCommand) {

        if (pssCommand.isRegistered()) {
            return null;
        }

        ICommand iCommand = new ICommand() {
            @Override
            public String getCommandName() {
                return pssCommand.getName();
            }

            @Override
            public String getCommandUsage(ICommandSender sender) {
                return pssCommand.getDescription();
            }

            @Override
            public List<String> getCommandAliases() {
                return pssCommand.getAliases();
            }

            @Override
            public void processCommand(ICommandSender sender, String[] args) throws CommandException {
                pssCommand.runRunnable(sender, args);
            }

            @Override
            public boolean canCommandSenderUseCommand(ICommandSender sender) {
                return true;
            }

            @Override
            public List<String> addTabCompletionOptions(ICommandSender sender, String[] args, BlockPos pos) {
                return new ArrayList<>();
            }

            @Override
            public boolean isUsernameIndex(String[] args, int index) {
                return false;
            }

            @Override
            public int compareTo(@NotNull ICommand o) {
                return this.getCommandName().compareTo(o.getCommandName());
            }
        };
        pssCommand.setICommand(iCommand);
        ClientCommandHandler.instance.registerCommand(iCommand);
        commandList.put(pssCommand.getName(), pssCommand);

        return iCommand;
    }

    public PSSCommand getCommand(String commandName) {
        return commandList.get(commandName);
    }

}

File: src/main/kotlin/me/partlysanestudios/partlysaneskies/commands/CommandManager.kt
//
// Written by Su386.
// See LICENSE for copyright and license notices.
//
package me.partlysanestudios.partlysaneskies.commands

import net.minecraft.command.CommandException
import net.minecraft.command.ICommand
import net.minecraft.command.ICommandSender
import net.minecraft.util.BlockPos
import net.minecraftforge.client.ClientCommandHandler

object CommandManager {
    var commandList = HashMap<String, PSSCommand>()

    fun getCommand(commandName: String): PSSCommand? = commandList[commandName]

    fun registerCommand(pssCommand: PSSCommand): ICommand? {
        if (pssCommand.isRegistered) return null
        val iCommand: ICommand =
            object : ICommand {
                override fun getCommandName(): String = pssCommand.name

                override fun getCommandUsage(sender: ICommandSender): String = pssCommand.description

                override fun getCommandAliases(): List<String> = pssCommand.aliases

                @Throws(CommandException::class)
                override fun processCommand(sender: ICommandSender, args: Array<String>) {
                    pssCommand.runRunnable(args)
                }

                override fun canCommandSenderUseCommand(sender: ICommandSender): Boolean = true

                override fun addTabCompletionOptions(sender: ICommandSender, args: Array<String>, pos: BlockPos): List<String> = ArrayList()

                override fun isUsernameIndex(args: Array<String>, index: Int): Boolean = false

                override fun compareTo(other: ICommand): Int = commandName.compareTo(other.commandName)
            }
        pssCommand.iCommand = iCommand
        ClientCommandHandler.instance.registerCommand(iCommand)
        pssCommand.isRegistered = true
        commandList[pssCommand.name] = pssCommand
        return iCommand
    }
}
